website/process_notebooks.py

```diff
@@ -121,56 +164,166 @@ def skip_reason_or_none_if_ok(notebook: Path) -> typing.Optional[str]:
+        intermediate_notebook.unlink()
 
-    if dry_run:
-        return colored(f"Would process {src_notebook.name}", "green")
+        if "extra_files_to_copy" in front_matter:
+            for file in front_matter["extra_files_to_copy"]:
+                (dest_dir / file).unlink()
 
-    # Copy notebook to target dir
-    # The reason we copy the notebook is that quarto does not support rendering from a different directory
-    shutil.copy(src_notebook, intermediate_notebook)
+        # Post process the file
+        post_process_mdx(target_mdx_file)
+    else:
+        target_mdx_file = src_notebook.with_suffix(".mdx")
 
-    # Check if another file has to be copied too
-    # Solely added for the purpose of agent_library_example.json
-    front_matter = extract_yaml_from_notebook(src_notebook)
-    # Should not be none at this point as we have already done the same checks as in extract_yaml_from_notebook
-    assert front_matter is not None, f"Front matter is None for {src_notebook.name}"
-    if "extra_files_to_copy" in front_matter:
-        for file in front_matter["extra_files_to_copy"]:
-            shutil.copy(src_notebook.parent / file, dest_dir / file)
-
-    # Capture output
-    result = subprocess.run(
-        [quarto_bin, "render", intermediate_notebook], stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True
-    )
-    if result.returncode != 0:
-        return colored(f"Failed to render {intermediate_notebook}", "red") + f"\n{result.stderr}" + f"\n{result.stdout}"
-
-    # Unlink intermediate files
-    intermediate_notebook.unlink()
+        # If the intermediate_notebook already exists, check if it is newer than the source file
+        if target_mdx_file.exists():
+            if target_mdx_file.stat().st_mtime > src_notebook.stat().st_mtime:
+                return colored(f"Skipping {src_notebook.name}, as target file is newer", "blue")
 
-    if "extra_files_to_copy" in front_matter:
-        for file in front_matter["extra_files_to_copy"]:
-            (dest_dir / file).unlink()
+        if dry_run:
+            return colored(f"Would process {src_notebook.name}", "green")
 
-    # Post process the file
-    post_process_mdx(target_mdx_file)
+        result = subprocess.run(
+            [quarto_bin, "render", src_notebook], stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True
+        )
+        if result.returncode != 0:
+            return colored(f"Failed to render {src_notebook}", "red") + f"\n{result.stderr}" + f"\n{result.stdout}"
 
     return colored(f"Processed {src_notebook.name}", "green")
 
 
+# Notebook execution based on nbmake: https://github.com/treebeardtech/nbmakes
+@dataclass
+class NotebookError:
+    error_name: str
+    error_value: Optional[str]
+    traceback: str
+    cell_source: str
+
+
+@dataclass
+class NotebookSkip:
+    reason: str
+
+
+NB_VERSION = 4
+
+
+def test_notebook(notebook_path: Path, timeout: int = 300) -> Tuple[Path, Optional[Union[NotebookError, NotebookSkip]]]:
```

Comment:
Could you document in notebook contribution guide about these options? Also add a link in Contribute.md.

Reply:
Updated